lib/Echidna/Output/Source.hs

```diff
@@ -0,0 +1,123 @@
+  in unlines $ map snd $ concatMap (\(f,vls) ->
+    (mempty, findFile f) :                                                   -- Add a header for each source file to show its complete path
+      filterLines covLines (map ((findFile f,) . decodeUtf8) $ V.toList vls) -- Show the source code for each file with its covered line.
+                                   ) allLines
+
+-- | Mark one particular line, from a list of lines, keeping the order of them 
+markLine :: Int -> TxResult -> FilePathText -> [(FilePathText, Text)] -> [(FilePathText, Text)] 
+markLine n r cf ls = case splitAt (n-1) ls of
+                        (xs, (f,y):ys) | f == cf -> xs ++ [(cf, markStringLine r $ unpack y)] ++ ys
+                        _                        -> map (\(f,y) -> (f, checkMarkers $ unpack y)) ls
+
+-- | Header preppend to each line
+markerHeader :: String
+markerHeader = "    |"
+
+-- | Add space for markers if necessary
+checkMarkers  :: String -> Text
+checkMarkers l@(_:_:_:_:'|':_) = pack l
+checkMarkers l                 = pack $ markerHeader ++ l
+
+-- | Add a proper marker to a line, and convert it to Text
+markStringLine  :: TxResult -> String -> Text
+markStringLine r (' ': ' ': ' ': ' ': '|':l) = pack $ getMarker r : ' ' : ' ' : ' ': '|' : l
+markStringLine r (m1: ' ': ' ': ' ':  '|':l)  = pack $ case getMarker r of
+                                                        m | m1 == m -> m1 : ' ' : ' ' : ' ': '|' : l
+                                                        m           -> m1 :  m  : ' ' : ' ': '|' : l
+
+markStringLine r (m1: m2 : ' ': ' ': '|':l)  = pack $ case getMarker r of
+                                                        m | m1 == m -> m1 :  m2  : ' ' : ' ': '|' : l
+                                                        m | m2 == m -> m1 :  m2  : ' ' : ' ': '|' : l
+                                                        m           -> m1 :  m2  : m   : ' ': '|' : l
+
+
+markStringLine r (m1: m2 : m3 : ' ': '|':l)  = pack $ case getMarker r of
+                                                        m | m1 == m -> m1 :  m2  : m3 : ' ': '|' : l
+                                                        m | m2 == m -> m1 :  m2  : m3 : ' ': '|' : l
+                                                        m | m3 == m -> m1 :  m2  : m3 : ' ': '|' : l
+                                                        m           -> m1 :  m2  : m3 : m  : '|' : l
+
+markStringLine _ (_: _ : _ : _ : '|':_)  = error "impossible to add another marker"
+markStringLine r l = pack $ getMarker r : ' ' : ' ' : ' ': '|' : l
+
+-- | Select the proper marker, according to the result of the transaction
+getMarker :: TxResult -> Char
+getMarker Success       = '*'
+getMarker ErrorRevert   = 'r' 
+getMarker ErrorOutOfGas = 'o'
+getMarker _             = 'e'
+
+-- | Filter the lines per file, marking each line
+filterLines :: [Maybe (FilePathText, Int, TxResult)] -> [(FilePathText, Text)] -> [(FilePathText, Text)]
```

Comment:
can we move this function before `markLine` to be consistent and read more general functions first?

Reply:
Done.